Library/Homebrew/utils/io_selector.rb

```diff
@@ -0,0 +1,79 @@
+require "delegate"
+require "English"
+
+require "extend/io"
+
+module Utils
+  #
+  # The class `IOSelector` is a wrapper for `IO::select` with the
+  # added benefit that it spans the streams' lifetimes.
+  #
+  # The class accepts multiple IOs which must be open for reading.
+  # It then notifies the client as data becomes available
+  # per-stream.
+  #
+  # Its main use is to allow a client to read both `stdout` and
+  # `stderr` of a subprocess in a way that avoids buffer-related
+  # deadlocks.
+  #
+  # For a more in-depth explanation, see:
+  #   https://github.com/Homebrew/brew/pull/2466
+  #
+  class IOSelector < DelegateClass(Hash)
+    attr_reader :separator
+
+    alias all_streams keys
+    alias all_tags values
+    alias tag_of fetch
+
+    def self.each_line_from(streams = {},
+      separator = $INPUT_RECORD_SEPARATOR, &block)
```

Comment:
Can you indent this further in; this looks currently like it's a line in the function.